_posts/2021-10-24-g2p-advanced-mappings.md

```diff
@@ -0,0 +1,329 @@
+|---|---|---|---|
+| dd | ð | VOWEL | VOWEL |
+
+You can also use abbreviations like this in the G2P studio by writing them in the section at the bottom of the page titled 'Custom Abbreviations' They will be automatically applied to your custom rules above.
+
+{% picture g2p-abbs.png %}
+
+### Regular Expressions
+
+[Regular expressions](https://en.wikipedia.org/wiki/Regular_expression) are used ubiquitously in programming to define certain search patterns in text. In fact, this is how `g2p` rules work! They eventually get compiled into a regular expression too. For the most part, you can add regular expression syntax to your rules. So, suppose you wanted to write a rule that deleted word-final 's', you could write the following:
+
+| in | out | context_before | context_after |
+|---|---|---|---|
+| s |  | \S | \s\|$ |
+
+As you can see [in this handy regular expression cheatsheet](https://cheatography.com/davechild/cheat-sheets/regular-expressions/) our rule turns 's' into nothing if it is preceded by \S (not white space) and followed by either \s (white space) or $ (end of the text).
+
+Note: There are some 'gotchas' with writing regular expressions using g2p. This is a technical note, but if you're writing some complicated regular expressions and they're not working, don't hesitate to [raise an issue](https://github.com/roedoejet/g2p/issues/new/choose). For example there are some active issues around edge cases where regular expressions and `g2p`'s custom [syntax for indices](#using-specific-indices) don't play nice together.
+
+### Using specific indices
+
+Even people familiar with using `g2p` might not be aware that one of its main features is that it preserves indices between input and output segments. Meaning that when we convert from something like 'kæt' to 'kʰæt' as in the first example, `g2p` knows that it's the 'k' that turned into the 'k' and 'ʰ' as seen below.
+
+{% picture k-to-kh.png %}
+
+The default interpretation of rule indices by `g2p` is that it matches the characters between the input and the output one-by-one in a given rule until it reaches the end of either one, then it matches any remaining characters in the longer part (input or output) to the last character of the shorter part. For example, compare the following examples where 'abc' is converted to 'ab' and glombs the excess input character onto the last output character:
+
+{% picture abc-ab.png %}
+
+and where 'ab' is converted to 'abc' and glombs the excess output character onto the last input character:
+
+{% picture ab-abc.png %}
+
+But what if - for some silly and imaginary reason - we want to show that it was actually 'a' that turned into 'c', and 'b' that turned into 'ab'? Well, we can use special `g2p` syntax for explicitly writing these indices. Instead of, 
+
+| in | out | context_before | context_after |
+|---|---|---|---|
+| ab | abc | |  |
+
+we can write
+
+| in | out | context_before | context_after |
+|---|---|---|---|
+| a{1}b{2} | ab{2}c{1} | |  |
```

Comment:
This example would be more intuitive to follow if you changed it to ab -> bca, so that the reader would immediately think, "right, I might want to use these indices if my rule does an inversion."